atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java

```diff
@@ -364,6 +365,50 @@ public DBType getDbType() {
         return connConfig.getDbType();
     }
 
+    @Override
+    public void setPassword(String newPassword) {
+        Preconditions.checkNotNull(newPassword, "password cannot be null");
+        State localState = state;
+        if (localState.type == StateType.ZERO) {
+            synchronized (this) {
+                // need to check again in case init just happened
+                localState = state;
+                if (localState.type == StateType.ZERO) {
```

Comment:
While I believe that this is fine (since the only `State` with type set to `ZERO` has all nulls, and so checking the type is equivalent to checking the whole object), the method `checkAndGetDataSourcePool` performs a similar operation in the sense that it switches against the possible states that we can be in, and notably it compares the whole object vs just the type.

To err on the side of caution, can we follow the structure used there? In other words:
1. use a switch statement
2. compare `state` against `localState` directly

Reply:
I thought about that, but that is actually suboptimal (and would result in needing to write duplicate code). When you examine the state again (while being synchronized) if the state happens to be the normal state, you would actually want to have the same behavior as what is outside of the synchronized block). It would also mean that you potentially need to add a third volatile read instead of what is written here where it is guaranteed a max of two volatile memory reads.

Reply:
I'd hope that the perf of this method isn't that critical (though now that you've written this more efficient impl it makes sense, of course)!